### PR TITLE
Add Architecture document

### LINKED ISSUE #111: Create an architecture document
Describe:

- File layout
- General theory on how this works

### PR DESCRIPTION
**Pull request by gordonwatts**
Describing the layout of the repo and the code

Fixes #111

### CODE REVIEW

File: ARCHITECTURE.md [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+At the most general level:
+
+1. Client starts with a `func_adl` query represented as a python `ast.AST`. The query is rendered as a python ast by the `func_adl` library.
+1. Client creates a `atlas_xaod_executor` object (or `cms_aod_executor`)
+1. Client calls `apply_ast_transformation` method with the `ast`. This does `ast` -> `ast` transformations that simplify and combine `ast` elements.
+1. Client calls `write_cpp_files` method with the resulting `ast` from the previous step. The output directory where the C++ files can be written
+   need to be given to the method as well.
+1. The C++ files produced are ready to run on the input files.
+
+Though only part of the tests, you can see how the template files are run against a file in the `test/atlas/xaod/LocalFile.py` file, specifically the `execute_result_async` method. Given a `func_adl` ast, it calls `atlas_xaod_executor` as above, and then uses a `docker` image to run against some test files.
```
Comment (sudo-panda) [inline ARCHITECTURE.md:18]: The function has been moved to `tests/utils/base.py`

File: func_adl_xAOD/common/util_scope.py [hunk 2]
```diff
@@ -1,6 +1,7 @@
 # Scope related utilities
 # see https://stackoverflow.com/questions/33533148/how-do-i-specify-that-the-return-type-of-a-method-is-the-same-as-the-class-itsel
 # for info on this next line. Already looking forward to python 4...
+# TODO: #118 Move this into the cpplib folder
```
Comment (sudo-panda) [inline func_adl_xAOD/common/util_scope.py:4]: `cpplib` folder? That doesn't exist anymore

Comment (gordonwatts) [reply func_adl_xAOD/common/util_scope.py:4]: This line should have been deleted anyway - thanks!

File: ARCHITECTURE.md (hunk 1 above)
Comment (gordonwatts) [reply ARCHITECTURE.md:18]: Fixed in next commit

**Pull request merged**